File: kernel/src/device/class/char.c

// SPDX-FileCopyrightText: 2025 Julian Scheffers <julian@scheffers.net>
// SPDX-FileType: SOURCE
// SPDX-License-Identifier: MIT

#include "device/class/char.h"

#include "errno.h"



// Read bytes from the device.
errno_size_t device_char_read(device_char_t *device, void *rdata, size_t rdata_len) {
    mutex_acquire_shared(&device->base.driver_mtx, TIMESTAMP_US_MAX);
    driver_char_t const *driver = (void *)device->base.driver;
    errno_size_t         res    = driver->read(device, rdata, rdata_len);
    mutex_release_shared(&device->base.driver_mtx);
    return res;
}

// Write bytes to the device.
errno_size_t device_char_write(device_char_t *device, void const *wdata, size_t wdata_len) {
    mutex_acquire_shared(&device->base.driver_mtx, TIMESTAMP_US_MAX);
    driver_char_t const *driver = (void *)device->base.driver;
    errno_size_t         res    = driver->write(device, wdata, wdata_len);
    mutex_release_shared(&device->base.driver_mtx);
    return res;
}

File: kernel/src/device/class/char.rs
// SPDX-FileCopyrightText: 2025 Julian Scheffers <julian@scheffers.net>
// SPDX-FileType: SOURCE
// SPDX-License-Identifier: MIT

use alloc::boxed::Box;

use crate::{
    bindings::error::{EResult, Errno},
    device::{BaseDriver, Device},
};

pub struct CharSpecialty {
    pub(super) driver: Option<Box<dyn CharDriver>>,
}

/// Character device driver functions.
pub trait CharDriver: BaseDriver {
    /// Read bytes from the device.
    fn read(&self, device: &Device, rdata: &mut [u8]) -> EResult<usize>;
    /// Write bytes to the device.
    fn write(&self, device: &Device, wdata: &[u8]) -> EResult<usize>;
}

impl Device {
    /// Read bytes from the device.
    pub fn char_read(&self, rdata: &mut [u8]) -> EResult<usize> {
        let guard = self.state.lock_shared();
        let chardev = guard.specialty.as_char().ok_or(Errno::EBADF)?;
        let driver = chardev.driver.as_deref().ok_or(Errno::EAGAIN)?;
        driver.read(self, rdata)
    }

    /// Write bytes to the device.
    pub fn char_write(&self, wdata: &[u8]) -> EResult<usize> {
        let guard = self.state.lock_shared();
        let chardev = guard.specialty.as_char().ok_or(Errno::EBADF)?;
        let driver = chardev.driver.as_deref().ok_or(Errno::EAGAIN)?;
        driver.write(self, wdata)
    }
}
